### PR TITLE
Move semantics into VueJS components

### PR DESCRIPTION
**Pull request by lilyx**

### CODE REVIEW

File: src/components/trace-element.vue [hunk 1]
```diff
@@ -1,17 +1,20 @@
...
     </div>
     <div v-if="!isLeaf" ref="children"
          class="children" :class="{vbox: vbox}"
          :hidden="collapsed">
       <trace-element v-for="child in childrenToRender"
                      :id="child.id" :traceNode="child.traceNode" :context="child.context"
                      :currentLR="child.currentLR" :measureInputText="measureInputText"
-                     :isInVBox="child.isInVBox" :eventHandlers="eventHandlers">
+                     :isInVBox="child.isInVBox" :eventHandlers="eventHandlers"
+                     :showResult="hasResult">
```
Comment (pdubroy) [inline src/components/trace-element.vue]: I'm a bit confused here why we need both `showResult` and `hasResult`. Do they mean something different?

Also, do we need an extra guard here (e.g. `ohmEditor.options.enableSemanticsEditor`) to make sure that the semantics editor is not enabled in the visualizer-only mode?

File: src/components/trace-element.vue [hunk 2]
```diff
@@ -142,7 +147,9 @@
       context: {type: Object},
 
       eventHandlers: {type: Object},
-      currentLR: {type: Object}
+      currentLR: {type: Object},
+
+      showResult: {type: Boolean}
```
Comment (pdubroy) [inline src/components/trace-element.vue]: Can you add a comment explaining what `showResult` is?

File: src/semanticsEditorWrapper.js [hunk 3]
```diff
@@ -1,42 +1,25 @@
 /* eslint-env browser */
+/* global Vue */
```
Comment (pdubroy) [inline src/semanticsEditorWrapper.js]: You need `var Vue = require('vue')` now.

File: src/semanticsEditorWrapper.js [hunk 4]
```diff
@@ -46,151 +29,406 @@
...
-  function updateSuggestionHighlight(entry) {
-    var currentEntry = $('#suggestions .selected');
-    var selectedEntry;
-    if (entry === 'pre') {
-      selectedEntry = currentEntry.previousElementSibling || currentEntry;
-    } else if (entry === 'next') {
-      selectedEntry = currentEntry.nextElementSibling || currentEntry;
-    } else {
-      selectedEntry = domUtil.closestElementMatching('entry', entry);
+  Vue.component('action-addtion', {
```
Comment (pdubroy) [inline src/semanticsEditorWrapper.js]: Why is this here, in addition to being in its own .vue file?

File: src/semanticsEditorWrapper.js (hunk 3 above)
Comment (pdubroy) [inline src/semanticsEditorWrapper.js]: In fact, it looks like a lot of the code here is duplicated. Can this file actually be removed now?

File: src/components/action-addtion.vue [hunk 5]
```diff
@@ -0,0 +1,41 @@
+<template>
```
Comment (lilyx) [inline src/components/action-addtion.vue]: Rename the file into `add-action-control.vue`, as the file is actually handle the html element for adding a new action to the selected operation.

File: src/components/trace-element.vue (hunk 1 above)
Comment (lilyx) [inline src/components/trace-element.vue]: It's a little confusing, so I rename the `showResult` into `initallyHasResult`. It is actually indicating whether a node should have a result by the time it's created. It's necessary because we are doing the selective rendering, so not all the nodes are created by default. `hasResult` is the actual state of having a result container appended to the node, it's manually mutable as need to update it upon selecting/unselecting an operation.

I am a little confused about the extra guard you mentioned. Are you suggesting that we should add a guard for showing the semantics editors as a whole, or just the semantics results that attached to the nodes? If you're thinking about the semantics results, I am not sure if the extra guard is necessary, as users could actually unselected the operations to disable showing the semantics results.

File: src/semanticsEditorWrapper.js (hunk 3 above)
Comment (lilyx) [inline src/semanticsEditorWrapper.js]: Sorry, this file is not actually active any more. I'll removed it.

File: src/components/action-addtion.vue (hunk 5 above)
Comment (pdubroy) [inline src/components/action-addtion.vue]: 👍

File: src/components/trace-element.vue (hunk 1 above)
Comment (pdubroy) [inline src/components/trace-element.vue]: Re: naming -- thanks for the clarification, and I think the new naming makes things much more clear.

The guard I meant was about showing the semantics editors as a whole. But I need to think a bit more about how to do this, so I think what I might do is merge this PR into a separate branch for now, and then once I decide how I want to handle enabling/disabling semantics editing, I can do that and then merge the branch into master.

**Comment by pdubroy** [comment]
Thanks Meixian! This looks good, I don't think there are any other changes you need to make. I'm going to close this PR because I'd like to keep these changes on a separate branch until I have time to figure out how I want to handling enabling and disabling the semantics editing entirely.

Thanks for all your hard work on this!

**Comment by alexwarth** [comment]
Yes, thanks Meixian and Pat for all your hard work!!!
